### PR TITLE
add --force=seccomp

### LINKED ISSUE #1527: `seccomp(2)` filters to replace fakeroot
@mej suggested that it may be possible to use `seccomp(2)` filters rather than `fakeroot` or `libfakeroot` to implement `ch-image build --force`, e.g. via `SECCOMP_RET_ERRNO` or `SECCOMP_RET_USER_NOTIF`.

### PR DESCRIPTION
**Pull request by reidpr**
This PR implements a new, much simpler way to do `--force`. See `ch-image.rst` diff for how it works.

Couple review notes:

1. The syscall numbers table needs to be proofread (or, maybe someone has a way for me to get this info automatically).

2. Do we actually need the `#ifdef`s and `configure` switch? It didn't initially build using `musl-gcc`, but for reasons I think don't apply on distros actually based on musl.

Closes #1527.

Big thanks to @mej for suggesting this approach!

### CODE REVIEW

File: test/build/60_force.bats [hunk 1]
```diff
@@ -61,10 +42,44 @@ RUN ["apt-get", "install", "-y", "hello"]
...
-    [[ $(echo "$output" | grep -Fc 'init step 1: checking: $') -eq 1 ]]
-    [[ $(echo "$output" | grep -Fc 'init step 1: $') -eq 1 ]]
     [[ $(echo "$output" | grep -Fc 'RUN: new command:') -eq 2 ]]
-    [[ $output = *'init: already initialized'* ]]
-    [[ $output = *'--force: init OK & modified 2 RUN instructions'* ]]
+    [[ $output = *'--force=seccomp: modified 2 RUN instructions'* ]]
     [[ $output = *'grown in 4 instructions: tmpimg'* ]]
 }
+
+@test "${tag}: dpkg(8)" {
```
Comment (mphinney1100) [inline test/build/60_force.bats:50]: Why don't these test have any `status -eq 0` or checks output?

**Review by mphinney1100 — CHANGES_REQUESTED** [review]

File: doc/ch-image.rst [hunk 2]
```diff
@@ -647,10 +675,62 @@ are three basic steps:
...
+   know.
+
+Workaround mode :code:`seccomp` (default)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This mode uses the kernel’s :code:`seccomp(2)` system call filtering to
+intercept certain privileged system calls, do absolutely nothing, and return
+success to the program.
+
+The quashed system calls are: :code:`capset(2)`; :code:`chown(2)` and friends;
```
[the remark this answers is not in the document]
Comment (reidpr) [reply doc/ch-image.rst:700]: It is! I love it. https://en.wiktionary.org/wiki/quash

File: test/build/60_force.bats (hunk 1 above)
Comment (reidpr) [reply test/build/60_force.bats:50]: Any command not wrapped in `run` will fail the test if it fails (i.e., `[[ $status -eq 0 ]]` is implied). So here, we only care about whether the build succeeds or fails, not any output. Hence no need for `run`.

**Review by lucaudill — APPROVED** [review]
LGTM

**Review by mphinney1100 — APPROVED** [review]

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -202,6 +202,12 @@ jobs:
           make $ch_makej
           sudo install -D -m644 argp.h /usr/include/x86_64-linux-musl/argp.h
           sudo install -D -m755 libargp.a /usr/lib/x86_64-linux-musl/libargp.a
+          # musl-gcc does not install Linux headers. This is a bad kludge.
+          # See: https://www.openwall.com/lists/musl/2017/11/23/1
```
Comment (mej) [inline .github/workflows/main.yml:206]: For lines 207-210, my thoughts:
- This looks to me like it's going to cause problems for jailed/installrooted builds.  There's probably something I'm missing since this section also calls `sudo` unconditionally, so I'll try to dig a little deeper (time permitting).
- Since only 3 of the 4 symlinks are absolute (the last one being relative), I'm assuming thought and intent have gone into making this determination already, but it's still worth a mention.  Specifically, is the intent to point to content on the root filesystem no matter what, or would the Linux headers likely be installed with the same prefix as the `musl` stuff?  You likely already know this, but just in case, `sudo ln -s ../linux ../asm-generic ../x86_64-linux-gnu /usr/include/x86_64-linux-musl/`
- This may be too nit-picky, but I would recommend placing a trailing `/` at the end of `/usr/include/x86_64-linux-musl` on all 4 lines.  Reading over these lines, it wasn't immediately obvious (to me) that the symlink path was a pre-existing directory into which new symlinks with the same names were being dropped; placing the trailing `/` at the end makes it intuitively obvious that it's a directory and not a file/symlink.  Solely cosmetic and entirely subjective, so feel free to ignore! 😀

File: bin/ch_core.c [hunk 4]
```diff
@@ -410,6 +512,91 @@ void run_user_command(char *argv[], const char *initial_dir)
    Tf (0, "can't execve(2): %s", argv[0]);
 }
 
+/* Set up the fake-syscall seccomp(2) filter. This computes and installs a
```
Comment (mej) [inline bin/ch_core.c:515]: From the bottom of my grammar nerd heart, _**THANK YOU**_ for knowing the difference between "setup," "set-up," and "set up!"  🏆 🤣

File: configure.ac [hunk 5]
```diff
@@ -116,8 +116,28 @@ AC_ARG_ENABLE([test],
...
+     need_seccomp=no
+     msg_seccomp=no],
+  [''],   # option not specified
+    [want_seccomp=yes
+     need_seccomp=no
+     msg_seccomp='if tested working'],
+  [*],    # anything else
+    [AC_MSG_ERROR([invalid --with-seccomp arg: $with_seccomp])])
+
 AC_ARG_WITH([libsquashfuse],
```
Comment (mej) [inline configure.ac:138]: The patch hunk below for the `--with-libsquashfuse` help string modernization doesn't seem to be related to this overall PR; should it be moved to a different one, or did I overlook something?

File: doc/ch-image.rst [hunk 6]
```diff
@@ -627,11 +638,29 @@ This one is (ironically) :code:`apt-get` failing to drop privileges:
   E: seteuid 100 failed - seteuid (22: Invalid argument)
   E: setgroups 0 failed - setgroups (1: Operation not permitted)
 
-By default, nothing is done to avoid these problems, though :code:`ch-image`
-does try to detect if the workarounds could help. :code:`--force` activates
-the workarounds: :code:`ch-image` injects extra commands to intercept these
-system calls and fake a successful result, using :code:`fakeroot(1)`. There
-are three basic steps:
+Charliecloud provides two different workaround to avoid these problems. Both
```
Comment (mej) [inline doc/ch-image.rst:641]: "`workaround`" should be "`workarounds`" (plural) if you want to use that word, but I might suggest a synonym since you already use "workaround" frequently throughout this new text.  Maybe "`mechanisms`" or "`techniques`?"

File: doc/ch-image.rst [hunk 7]
```diff
@@ -647,10 +676,62 @@ are three basic steps:
...
+checks will fail, e.g., :code:`getuid(2)` after :code:`setuid(2)`.
+
+While this mode does not provide consistency, it does offer a hook to help
+prevent programs asking for consistency. For example, :code:`apt-get -o
+APT::Sandbox::User=root` will prevent :code:`apt-get` from attempting to drop
+privileges, which `it verifies
+<https://salsa.debian.org/apt-team/apt/-/blob/cacdb549/apt-pkg/contrib/fileutl.cc#L3343>`_,
+exiting with failure if the correct IDs are not found (which they won’t be
+under this approach). This can be expressed with
+:code:`--force-cmd=apt-get,-o,APT::Sandbox::User=root`, though this particular
```
Comment (mej) [inline doc/ch-image.rst:718]: I totally agree with @mphinney1100's previous (resolved) criticism above, in the help section for `--force-cmd`.  It didn't click with me what it was for...until I got here!  Once I understood `apt`'s validation procedure and required circumvention, it all suddenly made sense!

Might be helpful to add "`Consult the "Workaround mode seccomp" section below for details on why this can be useful (and may be necessary).`"  Maybe even a link to this section, if that's feasible/practical.

File: lib/build.py [hunk 8]
```diff
@@ -92,6 +88,28 @@ def main(cli_):
...
+      # convert cli.force_cmd to parsed dict
+      force_cmd = dict()
+      for line in cli.force_cmd:
+         (cmd, args) = force.force_cmd_parse(line)
+         force_cmd[cmd] = args
+      cli.force_cmd = force_cmd
+   ch.VERBOSE("force mode: %s" % cli.force)
+   if (cli.force == "seccomp"):
+      for (cmd, args) in cli.force_cmd.items():
+         ch.VERBOSE("force command: %s" % ch.argv_to_string([cmd] + args))
```
Comment (mej) [inline lib/build.py:107]: I know precisely 💩 about Python, so ignore this if I'm Smoking the Good Stuff...but the spacing/indentation, especially the `if` in line 108, looks wonky to me.  Was that intentional?

File: lib/build.py [hunk 9]
```diff
@@ -92,6 +88,28 @@ def main(cli_):
...
+   if (    cli.force == "seccomp"
+       and ch.cmd([ch.CH_BIN + "/ch-run", "--feature=seccomp"],
```
Comment (mej) [inline lib/build.py:109]: Okay, after a dozen or so read-throughs and some sleep, I think I see it now -- `cli.force` is preceded by spaces so that it lines up with `ch.cmd` and leaves room for the `and`, yes?  I'm not sure exactly why it threw me off nor quite how to eliminate the problem (that I've yet to identify...).  I do it differently, so maybe it's just my being unaccustomed to that style....

The only thing coming to mind is to maybe put a space before `ch.cmd` so that the `.` in `cli.` lines up with the one in `ch.`...maybe?  🤷

File: lib/force.py [hunk 10]
```diff
@@ -280,121 +281,119 @@
...
 
-def detect(image, force, no_force_detect):
-   f = None
-   if (no_force_detect):
-      ch.VERBOSE("not detecting --force config, per --no-force-detect")
+def new(image_path, force_mode, force_cmds):
+   """Return a new forcer object appropriate for image at image_path in mode
+      force_mode. If no such object can be found, exit with error."""
+   if (force_mode is None):
+      return Noop()
```
Comment (mej) [inline lib/force.py:297]: You should name it `Nope()` instead.  😉

File: lib/force.py [hunk 11]
```diff
@@ -280,121 +281,119 @@
...
-   return f
+      assert False, "unreachable code reached"
+
+def force_cmd_parse(text):
+   # 1. Split on “,” preceded by even number of backslashes.
+   #
+   # FIXME: Said backslashes are removed in the split, so you can’t have a
+   # component with trailing backslashes. That seems rare so I’m not fixing
+   # for now.
+   args = re.split(r"(?<!\\)(?:\\\\)*,", text)
```
Comment (mej) [inline lib/force.py:311]: I'm not sure I'm understanding the `FIXME` correctly, but would something like this help?
```python
>>> x = re.compile(r"((?<!\\)(?:\\\\)*,)")
>>> m = x.split(r"echo,I,can't,run,my,command,apt\,fix\,all-the-things\,--please,on,my,C:\\,drive")
>>> print(m)
['echo', ',', 'I', ',', "can't", ',', 'run', ',', 'my', ',', 'command', ',', 'apt\\,fix\\,all-the-things\\,--please', ',', 'on', ',', 'my', ',', 'C:', '\\\\,', 'drive']

>>> print(m[0::2])
['echo', 'I', "can't", 'run', 'my', 'command', 'apt\\,fix\\,all-the-things\\,--please', 'on', 'my', 'C:', 'drive']
>>> print(m[1::2])
[',', ',', ',', ',', ',', ',', ',', ',', ',', '\\\\,']
```
Parenthesizing the whole RE like that saves each delimiter string in between the fields themselves, as you can see in `m` above, and it's relatively painless to pull out/separate the two individual lists.

File: test/force-auto.py.in [hunk 12]
```diff
@@ -14,23 +14,23 @@
...
+# This would appear to yield 3×2×4 = 24 tests per distribution. However:
 #
-#   1. We only try pre-prepared images for "really need" commands with --force
-#      given, to save time, so it's at most 9 potential tests.
+#   1. We only try pre-prepared images for “really need” commands with --force
+#      given, to save time, so it’s at most 9 potential tests.
 #
-#   2. Some distros don't have any pre-preparation step, so that test doesn't
-#      make sense.
+#   2. The pre-preparation set doesn’t make sense for some distros or for
```
Comment (mej) [inline test/force-auto.py.in:30]: Should this say "`pre-preparation step`" rather than "`pre-preparation set`?"

File: test/force-auto.py.in [hunk 13]
```diff
@@ -14,23 +14,23 @@
...
+#      --force=seccomp.
 #
-#   3. Some distros don't have an "apparently need" command determined.
+#   3. Some distros don’t have an “apparently need” command determined.
```
Comment (mej) [inline test/force-auto.py.in:33]: This sounds awkward to me.  What about:  "`We've not yet determined an "apparently need" command for some distros.`?"  or "`The "apparently need" command for some distros has yet to be determined.`?"

File: test/force-auto.py.in [hunk 14]
```diff
@@ -303,6 +299,7 @@ class T_Arch_Latest(Test):
    config = "arch"
    base = "archlinux:latest"
    arch_excludes = ["aarch64", "ppc64le"]  # base only amd64
+   force_excludes = ["seccomp"]  # issue #FIXME
```
Comment (mej) [inline test/force-auto.py.in:302]: Did you want to open this issue and fix this `FIXME` comment prior to getting this PR merged in?

**Review by mej — APPROVED** [review]
LGTM overall.  I left a few suggestions here and there, but there aren't really any showstoppers.  I'll go ahead and approve now in case you want to address my comments later (or ignore them completely), but I'm also happy to dialogue about them and later re-approve as needed.

File: .github/workflows/main.yml (hunk 3 above)
Comment (reidpr) [reply .github/workflows/main.yml:206]: Indeed, I agree this is really ugly. However the only place it’s used is to set up the GitHub CI virtual machine, so I felt mostly OK about it.

File: configure.ac (hunk 5 above)
Comment (reidpr) [reply configure.ac:138]: Yeah, that unrelated fix is a little sloppy. ;)

File: lib/build.py (hunk 8 above)
Comment (reidpr) [reply lib/build.py:107]: Yes it is intentional, and most Python programmers do think I'm smoking something. (See also the 3-space indents.)

The reasoning is to make the two clauses and the `and` operator appear as three separate things. I like to imagine moving the `and` up half a line.

File: lib/force.py (hunk 10 above)
Comment (reidpr) [reply lib/force.py:297]: Correct.

File: lib/force.py (hunk 11 above)
Comment (reidpr) [reply lib/force.py:311]: Yeah that's not quite right (I think the desired output is `['echo', 'I', "can't", 'run', 'my', 'command', 'apt,fix,all-the-things,--please', 'on', 'my', 'C:\drive']`) but I'm still having a hard time grokking exactly what's up. The problem had to do with strings that terminate in a backslash.
